### ISSUE TITLE
Randomly marks objects as not cacheable

### ISSUE THREAD

**Issue by mqudsi** [pagespeed/ngx_pagespeed#963]
I have a problem with ngx_pagespeed seemingly-randomly marking resources as uncacheable. For example, this is a recent log entry:

    [Wed, 06 May 2015 10:15:12 GMT] [Info] [16980] HTTPCache key=https://neosmart.net/forums/css.php?css=xenforo,form,public&style=2&dir=LTR&d=1430516531 fragment=neosmart.net: remembering not-cacheable status for 175 seconds.

The headers for this requested resource:

````
C:\Users\Mahmoud>curl -I "https://neosmart.net/forums/css.php?css=xenforo,form,public&style=2&dir=LTR&d=1430516531"
HTTP/1.1 200 OK
Server: nginx/1.9.0
Date: Wed, 06 May 2015 10:16:13 GMT
Content-Type: text/css; charset=utf-8
Connection: keep-alive
Vary: Accept-Encoding
x-powered-by: PHP/5.6.5
expires: Wed, 01 Jan 2020 00:00:00 GMT
last-modified: Thu, 01 Jan 1970 00:00:00 GMT
cache-control: public
Proxied-To: LiteSpeed
````

I can see nothing in there that would prevent pagespeed from caching this resource. I see the `Vary: Accept-Encoding` header, but as I understand it, ngx_pagespeed will still cache while honoring the `Vary` header to serve different content for different values of `Accept-Encoding`? The `Date` is correct, the `last-modified` is obviously wrong but it shouldn't matter (unless it does) because the `expires` is set way in the future and `cache-control` is set to `public`. I wouldn't imagine it to be a case-sensitivity issue, either...

On a slightly-unrelated note, the page which requested this resource is subsequently served with the rewritten-but-uncacheable resource present in the modified HTML source, meaning that pageload is "broken".

e.g. the page has been rewritten to reference `https://neosmart.net/forums/css.php,qcss==xenforo,,form,,public,astyle==2,adir==LTR,ad==1430516531+css.php,qcss==discussion_list,,inline_mod,,moderator_bar,astyle==2,adir==LTR,ad==1430516531.pagespeed.cc.OLotRAl67gk-h2DyEDtD.css` instead of `https://neosmart.net/forums/css.php?css=xenforo,form,public&style=2&dir=LTR&d=1430516531`

Subsequent refreshes of the same page don't have this issue as ngx_pagespeed remembers the (false?) not-cacheable status for the CSS resource and does not attempt a rewrite. Shouldn't nginx "un-rewrite" the HTML to serve the original resource instead of the one that it realizes won't work or is this by-design because the resource is never tested until it's actually called?

**Comment by jeffkaufman** [comment]
I made a test page for this: http://www.jefftk.com/neosmart

It has PageSpeed configuration:

    pagespeed Domain https://neosmart.net;
    pagespeed FetchHttps enable;

It needs the `Domain` command because normally it would leave resources on other sites alone, and it needs `FetchHttps` because your site is on HTTPS.  You  would only need `FetchHttps`; do you have that on?  Or a `MapOriginDomain` instead?

My html references that file:

    <link rel=stylesheet href="https://neosmart.net/forums/css.php?css=xenforo,form,public&style=2&dir=LTR&d=1430516531">

On the first request I see background optimization but the css file link stays the same, as expected.  I don't see any "remembering not-cacheable status" messages for the css file in my log.  On the second request I get the css file rewritten to `https://neosmart.net/forums/A.css.php,qcss=xenforo,,form,,public,astyle=2,adir=LTR,ad=1430516531.pagespeed.cf.GAfUMbY8UR.css`, which isn't the same as the url you got, I think because your url also involves css-combining because you have multiple css files on your site and my test doesn't.

When testing that url, I get a 404, which is surprising.  Since you're running PageSpeed it should interpret that url instead of 404ing.

Could you paste your PageSpeed configuration directives?

**Comment by mqudsi** [comment]
I do have `FetchHttps enable` on, but your comment made me realize a possible cause for part of the problem - I still had `MapOriginDomain` set to load HTTPS content from the HTTP domain. At the same time, I have HTTP calls set to issue a HTTP 307 redirect to the HTTPS resource (because of a recent switch to all HTTPS).

Disabling the MapOriginDomain (well, reversing it, actually) takes care of that.

Now my original rewritten resource is served OK (`https://neosmart.net/forums/css.php,qcss==xenforo,,form,,public,astyle==2,adir==LTR,ad==1430516531+css.php,qcss==discussion_list,,inline_mod,,moderator_bar,astyle==2,adir==LTR,ad==1430516531.pagespeed.cc.OLotRAl67gk-h2DyEDtD.css`), but your rewritten one to `A.css.php` still isn't (`https://neosmart.net/forums/A.css.php,qcss=xenforo,,form,,public,astyle=2,adir=LTR,ad=1430516531.pagespeed.cf.GAfUMbY8UR.css`)

Physically, `css.php` exists while `A.css.php` does not. I do test for the existence of a file before proxying to the PHP backend for security reasons, but I'd imagine both the `css.php,qcss....` and the `A.css.php` requests equally do not pass through to the fastcgi proxy since neither exist?

Also, I don't know if this explains why the URL is being rewritten in the original request instead of being backgrounded for future requests only?

My pagespeed configuration (including the commented-out, earlier, incorrect `MapOriginDomain`):

````
	pagespeed FileCachePath /var/cache/nginx/pagespeed_temp/;
	pagespeed MessageBufferSize 100000;
	pagespeed InstallCrashHandler on;
	pagespeed EnableCachePurge on;
	pagespeed PurgeMethod PURGE;
	pagespeed ForceCaching on;
	#native fetcher doesn't support HTTPS content 
	pagespeed UseNativeFetcher off;
	pagespeed FetchHttps enable;
````


````
include conf/pagespeed_libraries.conf;

location ~ "\.pagespeed\.([a-z]\.)?[a-z]{2}\.[^.]{10}\.[^.]+" { add_header "" ""; }
location ~ "^/ngx_pagespeed_static/" { }
location ~ "^/ngx_pagespeed_beacon$" { }
pagespeed Statistics on;
pagespeed StatisticsLogging on;
pagespeed LogDir /var/log/nginx/pagespeed;
pagespeed AdminPath /pagespeed_admin;

location ~ ^/pagespeed_admin {
  auth_basic_user_file /etc/nginx/htpasswd;
}

pagespeed on;
pagespeed RewriteDeadlinePerFlushMs 5;
pagespeed RewriteLevel CoreFilters;

pagespeed EnableFilters trim_urls,pedantic,extend_cache,prioritize_critical_css,dedup_inlined_images,inline_images,insert_dns_prefetch,resize_mobile_images,canonicalize_javascript_libraries,rewrite_images;
pagespeed Domain https://neosmart.net;
pagespeed Domain http://neosmart.net;

pagespeed EnableFilters rewrite_domains;
pagespeed MapRewriteDomain https://neosmart.net http://neosmart.net;
#pagespeed MapOriginDomain http://neosmart.net https://neosmart.net;
pagespeed MapOriginDomain https://neosmart.net http://neosmart.net;

pagespeed InPlaceResourceOptimization on;
pagespeed DisableRewriteOnNoTransform on;
pagespeed InlineResourcesWithoutExplicitAuthorization Script,Stylesheet;
pagespeed FetchHttps enable,allow_self_signed;
pagespeed SslCertDirectory /etc/ssl/certs;
pagespeed UrlSigningKey ****redacted*****
````

**Comment by jeffkaufman** [comment]
It looks like the reason my css request failed was that you have `UrlSigningKey` set, which should have exactly this effect: blocking external attempts to create `.pagespeed.` resources.

So it looks like you're ok now?  And the problem was the `MapOriginDomain` combined with the redirect to https?

**Comment by mqudsi** [comment]
It depends - should I still have had the broken, rewritten resource in the HTML output? And should ngx_pagespeed have followed the 307 redirect to HTTPS?

Thanks!

**Comment by jeffkaufman** [comment]
> should I still have had the broken, rewritten resource in the HTML output?

No, you're right, that shouldn't happen.

Were you experimenting with different configuration settings at the time?  Is it possible that you had one set of configuration options that allowed the resource be fetched and optimized, then configuration options changed so it was no longer possible to fetch but the optimized resource was in cache, and then the optimized resource got evicted from cache?

> should ngx_pagespeed have followed the 307 redirect to HTTPS?

The point of the `MapOriginDomain` setup is so PageSpeed can fetch resources without having to support HTTPS.  So redirecting back to HTTPS wouldn't fix anything.

Separately, PageSpeed doesn't follow redirects on resources, though this is something we'd like to handle at some point. https://github.com/pagespeed/mod_pagespeed/issues/262

**Comment by mqudsi** [comment]
No, the only recent configuration change was fixing MapOriginDomain during/after our conversation. I also always hard clear my cache after each configuration change then kill and reload nginx entirely.
